src/posposet/poset.go

```diff
@@ -0,0 +1,363 @@
+package posposet
+
+import (
+	"sort"
+	"sync"
+)
+
+// Poset processes events to get consensus.
+type Poset struct {
+	store     *Store
+	state     *State
+	flagTable FlagTable
+	frames    map[uint64]*Frame
+
+	processingWg   sync.WaitGroup
+	processingDone chan struct{}
+
+	newEventsCh      chan *Event
+	incompleteEvents map[EventHash]*Event
+}
+
+// New creates Poset instance.
+func New(store *Store) *Poset {
+	const buffSize = 10
+
+	p := &Poset{
+		store:  store,
+		frames: make(map[uint64]*Frame),
+
+		newEventsCh:      make(chan *Event, buffSize),
+		incompleteEvents: make(map[EventHash]*Event),
+	}
+
+	p.bootstrap()
+	return p
+}
+
+// Start starts events processing. It is not safe for concurrent use.
+func (p *Poset) Start() {
+	if p.processingDone != nil {
+		return
+	}
+	p.processingDone = make(chan struct{})
+	p.processingWg.Add(1)
+	go func() {
+		defer p.processingWg.Done()
+		//log.Debug("Start of events processing ...")
+		for {
+			select {
+			case <-p.processingDone:
+				//log.Debug("Stop of events processing ...")
+				return
+			case e := <-p.newEventsCh:
+				p.onNewEvent(e)
+			}
+		}
+	}()
+}
+
+// Stop stops events processing. It is not safe for concurrent use.
+func (p *Poset) Stop() {
+	if p.processingDone == nil {
+		return
+	}
+	close(p.processingDone)
+	p.processingWg.Wait()
+	p.processingDone = nil
+}
+
+// PushEvent takes event into processing. Event order doesn't matter.
+func (p *Poset) PushEvent(e Event) {
+	e.parents = nil
+	p.newEventsCh <- &e
+}
+
+// onNewEvent runs consensus calc from new event. It is not safe for concurrent use.
+func (p *Poset) onNewEvent(e *Event) {
+	if p.store.HasEvent(e.Hash()) {
+		log.WithField("event", e).Warnf("Event had received already")
+		return
+	}
+
+	if e.parents == nil {
+		e.parents = make(map[EventHash]*Event, len(e.Parents))
+		for hash := range e.Parents {
+			e.parents[hash] = nil
+		}
+	}
+
+	nodes := newParentsValidator(e)
+	ltime := newLamportTimeValidator(e)
+
+	// fill event's parents index or hold it as incompleted
+	for hash := range e.Parents {
+		if hash.IsZero() {
+			// first event of node
+			if !nodes.IsParentUnique(e.Creator) {
```

Comment:
Instead of checking every single event for being the first it is better to create leaf events at the time of poset creation, which are roots, clothos and atropos of the frame 0, then you have base for recursion and every event with number of parents less that k would be invalid (according to lachesis paper).
See how it is implemented in current poset: 
https://github.com/Fantom-foundation/go-lachesis/blob/95e43ed2e6c6532b2bf4be8e15ca16b6785eeb25/src/poset/poset.go#L113

Reply:
There no static node list in PoS-network, so it is impossible to create leaf events at the time of poset creation. Checking every single event for being the first is cheaper than the same checking and leaf creating.

Reply:
Dynamic node list is not a problem. When you instantiate a new network you know at least two peers, otherwise it would not be a network. When a node joins existing network it should know at least one existing node of the network to start communicate with. So you always know at least two nodes and thus can create leaf events for those. When a node becomes aware of a new node on the network it can create leaf event for it

Reply:
@Maxime2, i see you did not catch my answer essence. Again: there is no **static** node list => need to detect leaf-ref all the time, not at genesis only => there is no virtual event profit, but virtual event spends resources for creating and storing.
And like I said before the probable possibility to improve code is not a reason for PR rejecting. It is a reason to suggest your own PR.